vkubelet/podcontroller.go

```diff
@@ -0,0 +1,339 @@
+	maxRetries = 20
+)
+
+// PodController is the controller implementation for Pod resources.
+type PodController struct {
+	// server is the instance to which this controller belongs.
+	server *Server
+	// podsInformer is an informer for Pod resources.
+	podsInformer v1.PodInformer
+	// podsLister is able to list/get Pod resources from a shared informer's store.
+	podsLister corev1listers.PodLister
+
+	// workqueue is a rate limited work queue.
+	// This is used to queue work to be processed instead of performing it as soon as a change happens.
+	// This means we can ensure we only process a fixed amount of resources at a time, and makes it easy to ensure we are never processing the same item simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+	// recorder is an event recorder for recording Event resources to the Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// NewPodController returns a new instance of PodController.
+func NewPodController(server *Server) *PodController {
+	// Create an event broadcaster.
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(log.L.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: server.k8sClient.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "pod-controller"})
+
+	// Create an instance of PodController having a work queue that uses the rate limiter created above.
+	pc := &PodController{
+		server:       server,
+		podsInformer: server.podInformer,
+		podsLister:   server.podInformer.Lister(),
+		workqueue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pods"),
+		recorder:     recorder,
+	}
+
+	// Set up event handlers for when Pod resources change.
+	pc.podsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(pod interface{}) {
+			if key, err := cache.MetaNamespaceKeyFunc(pod); err != nil {
+				runtime.HandleError(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+		UpdateFunc: func(_, pod interface{}) {
+			if key, err := cache.MetaNamespaceKeyFunc(pod); err != nil {
+				runtime.HandleError(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+		DeleteFunc: func(pod interface{}) {
+			if key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(pod); err != nil {
+				runtime.HandleError(err)
+			} else {
+				pc.workqueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	// Return the instance of PodController back to the caller.
+	return pc
+}
+
+// Run will set up the event handlers for types we are interested in, as well as syncing informer caches and starting workers.
+// It will block until stopCh is closed, at which point it will shutdown the work queue and wait for workers to finish processing their current work items.
+func (pc *PodController) Run(ctx context.Context, threadiness int) error {
+	defer runtime.HandleCrash()
```

Comment:
Not sure I like this. We should just let panics bubble up, also seems like a todo item to remove this upstream.

Reply:
I'll look deeper into this.

Reply:
Wow, I think github hate my comment here. Damn...

So the panic bubbles up anyway, as we don't override [`ReallyCrash`](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/runtime/runtime.go#L56-L59).

Reply:
So no need for the call?

Reply:
Removed this in 66059f5.